### PR TITLE
Hyperparameters for datamodule

### LINKED ISSUE #3769: Hyperparameters for DataModules
## 🚀 Feature

Add a `save_hyperparameters` function to `LightningDataModule` and let the `Trainer` log it together with the model's hyperparameters.

### Motivation

DataModules are a great way to decouple the model code from the data it runs on.
People using datasets where pre-processing is not as fixed as in the common benchmarks need their DataModules to be configurable.
Examples would be:

* size of sliding windows
* maximum sequence length
* type of scaling (min-max, standardization, etc.)

Logging these hyperparameters is just as important for evaluating your model performance as the model's hyperparameters.
Therefore, they should be automatically logged by the trainer, too.

### Pitch

You are still searching for the perfect way to pre-process your data for maximum performance?
Keep all your efforts in order by logging the hyperparameters of your `LightningDataModule`.

### Alternatives

Right now, I manually define a hyperparameter dictionary as a member of my DataModule.
Afterward, I call `update` on the hparams property of my LightningModule.
This is pretty low-level code at the top-level of my script.

### Additional context

Code example for current solution:

```
class MyDataModule(pl.LightningDataModule):
    def __init__(self,
                 fd,
                 batch_size,
                 max_rul=125,
                 window_size=30,
                 percent_fail_runs=None,
                 percent_broken=None,
                 feature_select=None):

        ...

        self.hparams = {'fd': self.fd,
                        'batch_size': self.batch_size,
                        'window_size': self.window_size,
                        'max_rul': self.max_rul,
                        'percent_broken': self.percent_broken,
                        'percent_fail_runs': self.percent_fail_runs}

...

data = datasets.MyDataModule(...)
model.hparams.update(data.hparams)
```

### PR DESCRIPTION
**Pull request by tilman151**
## What does this PR do?

Fixes #3769 

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
@nateraw what do you think about this PR?

One thing that comes to my mind is, we need a follow up for the tuner, so that it can set the batch_size  /learning rate also on the hparams attribute of the datamodule, now that it has proper hyperparameters.

**Comment by tilman151** [comment]
@awaelchli is there anything else I have to do or do I just wait? Sorry for the dumb question, it's my first real PR.

**Comment by tilman151** [comment]
@awaelchli @edenlightning the issue linked to this PR was closed a while ago. Does that mean the PR will not be merged? Sorry again for this dumb question.

**Comment by SeanNaren** [comment]
hey @tilman151 let me chase up with core to see what's the status here and if I can get @nateraw to have a look!

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1799,42 +1713,21 @@ def to_torchscript(
...
-                line = re.sub(r"\s+", "", line, flags=re.UNICODE)
-                if ".hparams=" in line:
-                    return line.split("=")[1]
-        except Exception as e:
-            return "hparams"
-
-        return None
+        hparams = self._to_hparams_dict(datamodule.hparams)
+        if not hasattr(self, '_hparams'):
+            self._hparams = hparams
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1723]: I do not think that data shall overwrite model protected attributes...

**Comment by tilman151** [comment]
> seems one of the tests is breaking on drone, running a parity test in terms of speed for loading the MNIST dataset it seems? Could you have a look?

I failed to reproduce this on my machine after running the test three times. The log says `AssertionError: lightning [0.73370854 0.83868648] was slower than PT (threshold 0.82)`, so the test is only 0.01 seconds (I guess) over the threshold. @SeanNaren do you know how sensitive this threshold is? Could it be a fluke?

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (tilman151) [reply pytorch_lightning/core/lightning.py:1723]: Do you mean to use the property setter instead?

**Comment by Borda** [comment]
@tilman151 the test shall be fixed, can we rebase to get this done?

File: pytorch_lightning/core/lightning.py [hunk 2]
```diff
@@ -1799,42 +1713,21 @@ def to_torchscript(
...
+            self._hparams_initial = hparams
+        else:
+            colliding_keys = [key for key in hparams.keys() if key in self.hparams]
+            colliding_keys += [key for key in hparams.keys() if key in self.hparams_initial]
+            colliding_keys = set(colliding_keys)
+            if colliding_keys:
+                raise ValueError(f'Error while adding datamodule hparams: '
+                                 f'the keys {colliding_keys} are already present in the model hparams.')
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1731]:
```suggestion
                raise ValueError(
                    f'Error while adding datamodule hparams: the keys {colliding_keys} are already present in the model hparams.'
                )
```

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/core/lightning.py:1723]: ok, my bad, this is still Module itself...

File: pytorch_lightning/utilities/hparams_mixin.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+import re
+from argparse import Namespace
+from typing import Union, Any
+
+from pytorch_lightning.core.saving import PRIMITIVE_TYPES, ALLOWED_CONFIG_TYPES
+from pytorch_lightning.utilities import AttributeDict
+from pytorch_lightning.utilities.parsing import get_init_args
+
+
+class HyperparametersMixin:
```
Comment (Borda) [inline pytorch_lightning/utilities/hparams_mixin.py:25]: I am not very sure if we can introduce Mixins again, @williamFalcon

File: pytorch_lightning/utilities/hparams_mixin.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    def hparams(self, hp: Union[dict, Namespace, Any]):
+        hparams_assignment_name = self.__get_hparams_assignment_variable()
+        self._hparams_name = hparams_assignment_name
+        self._set_hparams(hp)
+        # this resolves case when user does not uses `save_hyperparameters` and do hard assignement in init
+        if not hasattr(self, "_hparams_initial"):
+            self._hparams_initial = copy.deepcopy(self._hparams)
+
+    def __get_hparams_assignment_variable(self):
+        """"""
```
Comment (Borda) [inline pytorch_lightning/utilities/hparams_mixin.py:146]:
```suggestion
```

**Review by Borda — CHANGES_REQUESTED** [review]
@PyTorchLightning/core-contributors shall be allowed to add Mixins again?

File: pytorch_lightning/core/lightning.py (hunk 2 above)
Comment (s-rog) [reply pytorch_lightning/core/lightning.py:1731]: Does this mean I can't have something like the same `input_dim` key in both model (for setting first layer dims) and datamodule (for setting resize dims) hparams?

Comment (tilman151) [reply pytorch_lightning/core/lightning.py]: Yeah, that's the idea. Previously I added a prefix to the data hyperparameters but it was getting convoluted with all the duplicate hparams. I could change it back again?

**Comment by norabelrose** [comment]
Are there any updates on this? I would use this feature and wouldn't mind doing a bit of work to help it get merged if needed.

**Comment by tilman151** [comment]
@norabelrose I'm still waiting on the core devs if they are okay with the new mixin. Until then I wasn't sure if it makes sense to update the PR for each an every new commit. If you want to help out there, feel free to do so 😉

File: pytorch_lightning/utilities/hparams_mixin.py (hunk 3 above)
Comment (tilman151) [reply pytorch_lightning/utilities/hparams_mixin.py:25]: @Borda Are there any updates on this?

Comment (Borda) [reply pytorch_lightning/utilities/hparams_mixin.py:25]: In general we shall avoid adding mixins which has some cross reference to other class attributes as they are uneasy to debug...
Cc: @tchaton

**Comment by Borda** [comment]
Let me ping them again =) @williamFalcon @tchaton
Cc: @PyTorchLightning/core-contributors

**Comment by carmocca** [comment]
What alternative do we have?

**Comment by Borda** [comment]
@PyTorchLightning/core-contributors mind review this PR?

**Comment by carmocca** [comment]
> @PyTorchLightning/core-contributors mind review this PR?

Marking as draft until conflicts resolved and tests are green, feel free to ping afterward.

It's okay to use a Mixin here. Anything else would require a huge refactor

**Comment by tilman151** [comment]
@carmocca that's nice to hear, at last. I'll try to resolve the conflicts asap.

**Comment by cn4750** [comment]
Is this PR still in development?

**Comment by tilman151** [comment]
@carmocca could you approve the test workflows? They should be all green now. If I wait longer, there will be conflicts again...

File: pytorch_lightning/core/lightning.py [hunk 5]
```diff
@@ -1952,18 +1863,26 @@ def to_torchscript(
 
         return torchscript_module
 
-    @property
-    def hparams(self) -> Union[AttributeDict, dict, Namespace]:
-        if not hasattr(self, "_hparams"):
-            self._hparams = AttributeDict()
-        return self._hparams
+    def add_datamodule_hparams(self, datamodule: LightningDataModule):
+        """Add the hparams of a LightningDataModule to the hparams and hparams_initial of this module."""
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:1867]: can you explain what this is and when this is used?
why can't the datamodule and lightning module maintain their own hparams?

please note the main value in `save_hyperparameters()` is that it captures the instantiation of the module, so we can do `LightningModule.load_from_checkpoint()`. 

but this does not exist for datamodules. 

in the assertions below I see

```python
    assert old_hparams == model.hparams
    assert old_hparams == model.hparams_initial
```

Isn't that problematic?

Comment (tilman151) [reply pytorch_lightning/core/lightning.py:1867]: The hyperparameters of the datamodule are added to the lightning module because only the hyperparameters of the lightning module are logged. The cited function is called inside the `attach_datamodule` function. This workflow is mainly inspired by the way I hacked the datamodule hyperparameters into lightning for my own projects.

Concerning the asserts: I had to add the hparams of the datamodule to `hparams_initial` because only the trainer only passes these to the logger. Considering that the added `hparams` are the initial hparams of the datamodule, I saw this as an acceptable solution. This problem would be solved by treating the hparams of the datamodule separately.

I could modify the PR to log the hparams of the lighting module and the datamodule separately. This would make it necessary to distinguish where the hparams came from in the logger. A simple solution would be adding a prefix when logging.

Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:1867]: Yes, I think the approach needs to be changed. It would be fine to merge the hparams at the logging stage but LightningModule should never contain hparams from a datamodule. Otherwise when we do

`LightningModule.load_from_checkpoint()` it would pass parameters to the init that don't exist. It means you would not be able to load your model and get a type error.

Comment (tilman151) [reply pytorch_lightning/core/lightning.py:1867]: Ok, I'll see if I find a way to do it. I hope it is okay to postpone making a '.load_from_checkpoint' function for datamodules, too.

Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:1867]: Thanks for your efforts. Yes that would not be required for this PR and could be discussed separately :)

So in summary, what we want here is:
- `model.save_hyperaparamters()` saves into `model.hparams`
- `datamodule.save_hyperparameters()` saves into `datamodule.hparams`
- no sharing of hyperparameters
- trainer grabs `model.hparams` and `datamodule.hparams` and logs the combined set

Comment (tilman151) [reply pytorch_lightning/core/lightning.py:1867]: Done. I adopted the behavior of the `add_datamodule_hparams` function that raised an error if there are duplicate hyperparameters.

**Comment by tshu-w** [comment]
Any update? love this PR 👀

File: pytorch_lightning/utilities/hparams_mixin.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    """Merge the hparams of a LightningModule and a LightningDataModule and return them.
+
+    If there is an overlap between the hparams in the LightningModule and in the
+    LightningDataModule an exception is raised.
+
+    Args:
+        lightning_hparams: the hyperparameters of a LightningModule
+        data_hparams: the hyperparameters of a LightningDataModule
+    """
+    colliding_keys = [key for key in data_hparams.keys() if key in lightning_hparams]
```
Comment (carmocca) [inline pytorch_lightning/utilities/hparams_mixin.py:171]: You can use the intersection of `dict_keys`

```suggestion
    colliding_keys = lightning_hparams.keys() & data_hparams.keys()
```

File: pytorch_lightning/utilities/hparams_mixin.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    if colliding_keys:
+        raise ValueError(
+            f'Error while merging hparams: the keys {colliding_keys} are present '
+            f'in both the LightningModules and LightningDataModules hparams.'
+        )
+    merged_hparams = copy.deepcopy(lightning_hparams)
+    merged_hparams.update(data_hparams)
+
+    return merged_hparams
```
Comment (carmocca) [inline pytorch_lightning/utilities/hparams_mixin.py:180]:
```suggestion
    return {**lightning_hparams, **data_hparams}
```

File: pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -903,11 +904,16 @@ def _run(self, model: 'pl.LightningModule') -> Optional[Union[_EVALUATE_OUTPUT,
...
         self.accelerator.pre_dispatch(self)
+        self._log_hyperparams()
 
+    def _log_hyperparams(self):
         # log hyper-parameters
         if self.logger is not None:
             # save exp to get started (this is where the first experiment logs are written)
-            self.logger.log_hyperparams(self.lightning_module.hparams_initial)
+            datamodule_hparams = self.datamodule.hparams_initial if self.datamodule is not None else {}
+            hparams_initial = merge_hparams(self.lightning_module.hparams_initial, datamodule_hparams)
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:914]: Not sure if having the `merge_hparams` is useful considering how simple it is to merge two dictionaries in python

File: tests/models/test_hparams.py [hunk 9]
```diff
@@ -41,6 +46,11 @@ def __init__(self, hparams):
         super().__init__()
         self.save_hyperparameters(hparams)
 
+    def training_step(self, batch, batch_idx):
```
Comment (carmocca) [inline tests/models/test_hparams.py:49]: Why is this necessary?

File: pytorch_lightning/utilities/hparams_mixin.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+                line = re.sub(r"\s+", "", line, flags=re.UNICODE)
+                if ".hparams=" in line:
+                    return line.split("=")[1]
+        except Exception:
+            return "hparams"
+
+        return None
+
+
+def merge_hparams(lightning_hparams: dict, data_hparams: dict) -> dict:
```
Comment (ananthsub) [inline pytorch_lightning/utilities/hparams_mixin.py:161]: could this take a variable list of dicts instead? what if in the future we want to have more components with hyperparameters? it'd also avoid the `lightning_hparams` reference here since this mixin technically doesn't know anything about the lightning module or datamodule

i agree with @carmocca though: if this merging is currently only called within the trainer, do we need this funciton on the mixin at all?

Comment (kaushikb11) [reply pytorch_lightning/utilities/hparams_mixin.py:161]: Yeah, I agree. Working on the changes right now, it doesn't need to be a separate functionality.

File: pytorch_lightning/utilities/hparams_mixin.py [hunk 11]
```diff
@@ -0,0 +1,180 @@
+# Copyright The PyTorch Lightning team.
```
Comment (ananthsub) [inline pytorch_lightning/utilities/hparams_mixin.py:1]: on directory structure, this is under utilities/hparams_mixin but is used by core/lightning. does it make sense for this to be part of core too? should we have a `core/mixins` subdirectory?

File: pytorch_lightning/trainer/trainer.py (hunk 8 above)
Comment (kaushikb11) [reply pytorch_lightning/trainer/trainer.py]: Removed & updated!

File: tests/models/test_hparams.py (hunk 9 above)
Comment (kaushikb11) [reply tests/models/test_hparams.py]: Removed & updated!

File: pytorch_lightning/utilities/hparams_mixin.py (hunk 10 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/hparams_mixin.py]: Updated!

File: pytorch_lightning/utilities/hparams_mixin.py (hunk 11 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/hparams_mixin.py:1]: On board with that, could move `device_dtype_mixin` & `hparams_mixin` to a separate directory in `core` in a follow up PR.

Comment (tchaton) [reply pytorch_lightning/utilities/hparams_mixin.py:1]: Yes, I like this idea too. Especially if we move to a more opt-in structure for the Lightning module.

File: pytorch_lightning/utilities/hparams_mixin.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    def hparams_initial(self) -> AttributeDict:
+        if not hasattr(self, "_hparams_initial"):
+            return AttributeDict()
+        # prevent any change
+        return copy.deepcopy(self._hparams_initial)
+
+    @hparams.setter
+    def hparams(self, hp: Union[dict, Namespace, Any]):
+        hparams_assignment_name = self.__get_hparams_assignment_variable()
+        self._hparams_name = hparams_assignment_name
```
Comment (awaelchli) [inline pytorch_lightning/utilities/hparams_mixin.py:137]: why are we adding back the setter? 
We specifically disallowed this. Users are not allowed to set the attribute directly. 
They should instead modify hparams by interacting with the `self.hparams` dict.

File: pytorch_lightning/utilities/hparams_mixin.py [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        self._set_hparams(hp)
+        # this resolves case when user does not uses `save_hyperparameters` and do hard assignement in init
+        if not hasattr(self, "_hparams_initial"):
+            self._hparams_initial = copy.deepcopy(self._hparams)
+
+    def __get_hparams_assignment_variable(self):
+        """
+        looks at the code of the class to figure out what the user named self.hparams
+        this only happens when the user explicitly sets self.hparams
+        """
```
Comment (awaelchli) [inline pytorch_lightning/utilities/hparams_mixin.py:147]: This is not allowed in Lightning today. If the user replaces the container we can't guarantee they can reload the model because it's ambiguous how to pass hyperparmeters in. 
Please reconsider! I believe this should be discussed outside of this PR.

**Review by awaelchli — COMMENTED** [review]
Great feature, very much looking forward to it! I tried it out and it works great!

I am convinced adding back the setter is the wrong decision. This will get us in the same trouble again that we had before we removed it. Please reconsider this change

File: pytorch_lightning/utilities/hparams_mixin.py (hunk 12 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/hparams_mixin.py]: Yup, removed & updated!

File: pytorch_lightning/utilities/hparams_mixin.py (hunk 13 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/hparams_mixin.py]: Removed & updated!

Comment (awaelchli) [reply pytorch_lightning/utilities/hparams_mixin.py]: amazing!

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/hparams_mixin.py (hunk 11 above)
Comment (Borda) [reply pytorch_lightning/utilities/hparams_mixin.py:1]: coming back to the root, I thought we want to away from mixins
cc: @awaelchli @justusschock

Comment (awaelchli) [reply pytorch_lightning/utilities/hparams_mixin.py:1]: my perspective has always been: we want to move away from these multiple-inheritance abuses in the trainer that are NOT mixins (can name the classes mixins as much as we want, they are not). 

the class introduced here is a true mixin, and it's use is justified IMO.

File: pytorch_lightning/utilities/hparams_mixin.py [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+    __jit_unused_properties__ = ["hparams", "hparams_initial"]
+
+    def save_hyperparameters(
+        self,
+        *args,
+        ignore: Optional[Union[Sequence[str], str]] = None,
+        frame: Optional[types.FrameType] = None
+    ) -> None:
+        """Save model arguments to ``hparams`` attribute.
```
Comment (ethanwharris) [inline pytorch_lightning/utilities/hparams_mixin.py:35]:
```suggestion
        """Save arguments to ``hparams`` attribute.
```

**Review by ethanwharris — APPROVED** [review]
LGTM, minor comment

**Comment by kaushikb11** [comment]
@Borda @tchaton Mind review?

**Review by tchaton — APPROVED** [review]

**Review by Borda — APPROVED** [review]
lgtm

**Comment by s-rog** [comment]
Should this be documented in the datamodule docs?
